.github/workflows/ci.yml

```diff
@@ -17,272 +17,346 @@
     steps:
       - uses: actions/checkout@master
-      - name: free disk space
-        run: |
-          sudo swapoff -a
-          sudo rm -f /swapfile
-          sudo apt clean
-          docker rmi $(docker image ls -aq)
-          df -h
-      - name: "Tests [MySQL5.7][Py3.7][core][kerberos]"
-        run: ./scripts/ci/ci_run_airflow_testing.sh --ignore=tests/providers
+      - uses: actions/setup-python@v1
+        with:
+          python-version: '3.x'
+      - name: "Free space"
+        run: ./scripts/ci/ci_free_space_on_ci.sh
+      - name: "Build CI image ${{ matrix.python-version }}"
+        run: ./scripts/ci/ci_prepare_image_on_ci.sh
+      - name: "Tests"
+        run: ./scripts/ci/ci_run_airflow_testing.sh
 
   requirements:
-    name: "Generate requirements"
+    timeout-minutes: 20
+    name: "Requirements: Py${{ matrix.python-version }}"
     runs-on: ubuntu-latest
-    needs: [statics, statics-tests]
     strategy:
       matrix:
         python-version: [3.6, 3.7]
+    env:
+      PYTHON_MAJOR_MINOR_VERSION: ${{ matrix.python-version }}
+      SHOW_GENERATE_REQUIREMENTS_INSTRUCTIONS: true
+      CI_JOB_TYPE: "Requirements"
     steps:
       - uses: actions/checkout@master
       - uses: actions/setup-python@v1
+      - name: "Build CI image ${{ matrix.python-version }}"
+        run: ./scripts/ci/ci_prepare_image_on_ci.sh
       - name: "Generate requirements"
-        env:
-          PYTHON_MAJOR_MINOR_VERSION: ${{ matrix.python-version }}
-          SHOW_GENERATE_REQUIREMENTS_INSTRUCTIONS: true
         run: ./scripts/ci/ci_generate_requirements.sh
 
-  build-image:
-    name: "Build production image"
+  push-prod-images-to-github-cache:
+    timeout-minutes: 60
+    name: "Push PROD images"
     runs-on: ubuntu-latest
-    needs: [statics, statics-tests]
+    needs:
+      - tests-sqlite
+      - tests-postgres
+      - tests-mysql
+      - requirements
+      - prepare-backport-packages
+      - build-prod-image
+      - docs
+    if: github.ref == 'refs/heads/master' && github.event_name != 'scheduled'
     strategy:
       matrix:
         python-version: [3.6, 3.7]
+    env:
+      PYTHON_MAJOR_MINOR_VERSION: ${{ matrix.python-version }}
+      CI_JOB_TYPE: "Prod image"
     steps:
       - uses: actions/checkout@master
-      - uses: actions/setup-python@v1
-      - name: "Build image"
-        env:
-          PYTHON_MAJOR_MINOR_VERSION: ${{ matrix.python-version }}
-        run: ./scripts/ci/ci_build_production_image.sh
-  prepare-backport-packages:
-    name: "Prepare & test backport packages"
+      - name: "Free space"
+        run: ./scripts/ci/ci_free_space_on_ci.sh
+      - name: "Build PROD images ${{ matrix.python-version }}"
+        run: ./scripts/ci/ci_build_production_images.sh
+      - name: "Push PROD images ${{ matrix.python-version }}"
+        run: ./scripts/ci/ci_push_production_images.sh
+
+  push-ci-images-to-github-cache:
+    timeout-minutes: 60
```

Comment:
Can it really take 60 mins to push to the GH registry? Or is this timeout cumulative including the time of upstream jobs?

Reply:
Nope. I can decrease it now after experimenting and seeing the result. will do.